source/common/upstream/cluster_manager_impl.cc

```diff
@@ -2037,6 +2037,17 @@ ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::httpConnPoolImp
 
   ConnPoolsContainer& container = *parent_.getHttpConnPoolsContainer(host, true);
 
+#ifdef ENVOY_ENABLE_QUIC
+  // If upstream is HTTP/3 and the downstream request is CONNECT-UDP, the maximum packet length
+  // should be adjusted to ensure we can fit QUIC packets inside DATAGRAM frames.
+  if (!upstream_protocols.empty() && upstream_protocols[0] == Http::Protocol::Http3 && context &&
+      context->downstreamHeaders() &&
+      Http::HeaderUtility::isConnectUdpRequest(*context->downstreamHeaders())) {
+    quic_info_ = Quic::createPersistentQuicInfoForCluster(
```

Comment:
Good point, so the `quic_info_` is always nullptr here in the current code path. A quic info object is assigned in `ProdClusterManagerFactory::allocateConnPool` if `quic_info_` is a nullptr (https://github.com/envoyproxy/envoy/blob/88543c9c37f389ff6d2650bb7aae211563ab0485/source/common/upstream/cluster_manager_impl.cc#L2259). Since this new code is exercised before the `ProdClusterManagerFactory::allocateConnPool`, the current code shouldn't have the dangling pointer issue.

I agree that this is still error-prone though. Another option I was thinking about is just to move this code to `ProdClusterManagerFactory::allocateConnPool` to at least make the instantiation of the quic_info object done in a single place. I was wondering how @danzh2010 would think about the current use of the `quic_info` in the current code, it seems to me that there's room for refactoring.

Reply:
I think the second question is more important here. My understanding is that the maximum packet length of a QUIC connection is set when it's initialized. This means we should configure all Envoy clusters to create QUIC connections with the larger maximum packet length if a CONNECT-UDP route is configured in the HTTP Connection Manager. Ideally, this configuration would be applied only to the clusters specifically listed under the CONNECT-UDP route, if possible. Does this sound correct, Ryan? I'd also like to hear @DavidSchinazi's thoughts on this.

Reply:
Another option might be to make this max packet length configurable in Envoy independently. This seems reasonable if we need to adjust this value when connections are created.

Reply:
Yeah I think this might be a much cleaner way to go. There is already a QuicProtocolOptions field which hsa, for example, `max_concurrent_streams` and `initial_stream_window_size` so putting `max_packet_size` there seems reasonable. https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/protocol.proto#envoy-v3-api-msg-config-core-v3-quicprotocoloptions

Reply:
Cool! I think this is the right way to go. I'll send a new commit to the PR soon, but @DavidSchinazi, let me know if you have any other thoughts.

Reply:
Yeah the max packet length needs to be set at QUIC connection setup time, so it can't depend on the method of the request that triggers the connection creation. We could do this based on whether connect-udp is supported at all, but having it be a separate config knob also SGTM

Reply:
I added a new `max_packet_length` QUIC protocol option. We don't update the `quic_info_` pointer anymore in the new code so the first question from Ryan and my answer to that is no longer relevant. Let me know any of you have other comments.